crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/bloom.rs

```diff
@@ -0,0 +1,174 @@
+        let mut bloom = bloomfilter::Bloom::new_for_fp_rate(agents_within_arc.len(), Self::TGT_FP);
+
+        for info in agents_within_arc {
+            let signed_at_ms = info.signed_at_ms;
+            // The key is the agent hash + the signed at.
+            let key = Arc::new(MetaOpKey::Agent(info.0.agent.clone(), signed_at_ms));
+            bloom.set(&key);
+        }
+        Ok(Some(bloom))
+    }
+
+    /// Generate a bloom filter of all ops.
+    /// - Ops are only included if they are within the common arc set.
+    /// - The bloom is `KitsuneOpHah`.
+    /// - Ops are only included from local agents that are within the common arc set.
+    /// - Only create the filter if there are any ops matching the above criteria.
+    /// No empty bloom filters.
+    /// - Bloom has a 1% chance of false positive (which will lead to agents not being sent back).
+    /// - Expect this function to complete in an average of 10 ms and worst case 100 ms.
+    pub(super) async fn generate_ops_blooms_for_time_window(
+        &self,
+        local_agents: &HashSet<Arc<KitsuneAgent>>,
+        common_arc_set: &Arc<DhtArcSet>,
+        mut search_time_window: Range<u64>,
+    ) -> KitsuneResult<Vec<TimedBloomFilter>> {
+        let mut results = Vec::new();
+        loop {
+            // Get the agents withing the arc set and filter by local.
+            let local_agents_within_arc_set: Vec<_> =
+                store::agents_within_arcset(&self.evt_sender, &self.space, common_arc_set.clone())
+                    .await?
+                    .into_iter()
+                    .filter(|(a, _)| local_agents.contains(a))
+                    .collect();
+
+            // Get the op hashes which fit within the common arc set from these local agents.
+            let result = store::all_op_hashes_within_arcset(
+                &self.evt_sender,
+                &self.space,
+                local_agents_within_arc_set.as_slice(),
+                common_arc_set,
+                search_time_window.clone(),
+                Self::UPPER_HASHES_BOUND,
+                true,
+            )
+            .await?;
+
+            // If there are none then don't create a bloom.
+            let (ops_within_common_arc, found_time_window) = match result {
+                Some(r) => r,
+                None => break,
+            };
+
+            let num_found = ops_within_common_arc.len();
+
+            // Create the bloom from the op hashes.
+            let mut bloom =
+                bloomfilter::Bloom::new_for_fp_rate(ops_within_common_arc.len(), Self::TGT_FP);
+
+            for hash in ops_within_common_arc {
+                bloom.set(&Arc::new(MetaOpKey::Op(hash)));
+            }
+
+            // If we found the maximum number of ops we can then
+            // there might still be more ops in the search window.
+            if num_found >= Self::UPPER_HASHES_BOUND
+                && found_time_window.end < search_time_window.end
+            {
```

Comment:
I don't understand why we have the multiple layers here, whereby we have an exponentially growing list of time ranges (now-1 hour ago, ... 1 month ago-beginning of time) - then, *within* those time ranges, we do this loop here to make multiple bloom filters...

Would it not be simpler to have the query have both an earliest time, and a max hash count. Then, the recent loop could set the earliest time to 1 hour ago, and the historical loop could set the earliest time to the beginning of time, and they could both return any number of bloom filters based on the max hash count?

Reply:
So for others reading this we discussed this and the time ranges are there for when we add hashing of time windows. It needs to be reworked to be relative to the unix epoch as well but this will work for now.